2019-03-Autumn/1_linked_lists/README.md

```diff
@@ -70,21 +107,77 @@ TODO :bug:
 
 TODO :bug:
 
-### 2. PROBLEM 2 TODO :bug:
-
-Source: TODO :bug:
-
-#### Naive/Simple Solution
-
-TODO :bug:
-
-#### Optimal Solution
-
-TODO :bug:
-
-#### Testing The Solutions OR Driver For Solution
-
-TODO :bug:
+### 2. Loop Detection
+
+Source: CTCI 2.8
+
+#### Solution for Detect Loop
+
+By using the *slow pointer and fast pointer* approach, we can guarantee that in 
+the case which there is a loop, the 2 pointers will meet at some node in the loop.
+
+```c++
+struct Node {
+    int data;
+    Node* next;
+};
+
+bool detectLoop(Node* head) {
+    Node* fast = head;
+    Node* slow = head;
+    while (fast != nullptr && fast->next != nullptr) {
+        fast = fast->next->next;
+        slow = slow->next;
+        if (fast == slow) {
+            return true;
+        }
+    }
+    // fast hit nullptr, no loop
+    return false;
+}
+```
+
+#### Solution for Start of Loop
+
+To find the start of the loop, we first do the same steps as the solution above.
+When the pointers collide, we move one of the pointers back to head, and move 
+both pointers again, but at the normal (one at the time) speed. When they 
+collide again, we have found the answer.
+
+
+[GeeksforGeeks](https://www.geeksforgeeks.org/detect-and-remove-loop-in-a-linked-list/)
+has a good explanation of why this works.
```

Comment:
So, this implementation is by far the most elegant, but it is also the most difficult to understand. Even the proof on G4G about why this works is long and involved and would not be easy to explain in 10 minutes or less.

I think that, to introduce the idea of finding the loop's starting point, it would be advisable to begin by counting the number of nodes in the cycle and moving pointers in lockstep with that interval (essentially Method no. 2 in the GeeksForGeeks writeup). That would be a good lead-up to the solution presented here, and would give the attendees some context about solving this problem in multiple ways.